src/main/java/in/projecteka/consentmanager/link/discovery/model/patient/request/PatientIdentifier.java

```diff
@@ -0,0 +1,17 @@
+package in.projecteka.consentmanager.link.discovery.model.patient.request;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import javax.validation.constraints.NotBlank;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class PatientIdentifier {
+    @NotBlank(message = "Identifier Type must be specified")
+    private PatientIdentifierType type;
+    @NotBlank(message = "Identifier Value must be specified")
```

Comment:
Problem with annotation is that error message what you get is not the same format what we specified in the contract.

Reply:
Then we will either externalise the error msg or specify the same. Do you want me to do that? But right now, there are so many apis where anything is accepted! This will at least stop that.

Reply:
Done. Extracted out constants. We can go and change the messages anytime in a single place.

Reply:
I did not mean the error message, if you hit the api by yourselves you will realise what I meant.

it's supposed to return 401 BAD REQUEST

```
{
"error": {
"code": "xxxx".
"message": "validation error message"
}
}
```

It's okay right now, but I think we should do what we are doing for user creation API.

Reply:
I believe there are different sorts of validations, and we should validate different types at different levels. If any input validations of schema, then I prefer to put javax validation. All such should result in BAD request. For that, the cleanest and standard way to do are JSR javax validations. Anything like a rule, like a pwd, ought to be a biz validations (because they are not schema validations) should be done in other layers, like you have done. (although I would have pushed that to the service, rather than controller). and for that, we have specific error codes.

Reply:
I'm not going to argue with you about having data validator at constructor (SignupValidator) and validating business rules where we need database acesss (validate whether user already exists) in the UserService is not bad idea. I understood people have difference of opinions. That's okay.

But, problem with using javax annotation validation is, with just relying on spring to return the error response, it's not interoperable. We define contract and client will try to respect the contract, and client could be in any language, not just java. Moreover, today the contract for spring with javax is something, tomorrow it might be something else.

If you hit create consent request, with some validation errors, you would response like this.

HTTPSTATUS - 500
body is 

```
{
    "timestamp": "2020-04-02T01:45:17.437+0000",
    "path": "/consent-requests",
    "status": 400,
    "error": "Bad Request",
    "message": "Validation failed for argument at index 0 in method: public reactor.core.publisher.Mono<in.projecteka.consentmanager.consent.model.response.RequestCreatedRepresentation> in.projecteka.consentmanager.consent.ConsentRequestController.requestConsent(in.projecteka.consentmanager.consent.model.request.ConsentRequest), with 2 error(s): [Field error in object 'consentRequest' on field 'consent.purpose': rejected value [null]; codes [NotNull.consentRequest.consent.purpose,NotNull.consent.purpose,NotNull.purpose,NotNull.in.projecteka.consentmanager.consent.model.ConsentPurpose,NotNull]; arguments [org.springframework.context.support.DefaultMessageSourceResolvable: codes [consentRequest.consent.purpose,consent.purpose]; arguments []; default message [consent.purpose]]; default message [Purpose is not specified.]] [Field error in object 'consentRequest' on field 'consent.callBackUrl': rejected value [null]; codes [NotNull.consentRequest.consent.callBackUrl,NotNull.consent.callBackUrl,NotNull.callBackUrl,NotNull.java.lang.String,NotNull]; arguments [org.springframework.context.support.DefaultMessageSourceResolvable: codes [consentRequest.consent.callBackUrl,consent.callBackUrl]; arguments []; default message [consent.callBackUrl]]; default message [Call back url not specified.]] ",
    "requestId": "26b1cba7-4",
    "errors": [
        {
            "codes": [
                "NotNull.consentRequest.consent.purpose",
                "NotNull.consent.purpose",
                "NotNull.purpose",
                "NotNull.in.projecteka.consentmanager.consent.model.ConsentPurpose",
                "NotNull"
            ],
            "arguments": [
                {
                    "codes": [
                        "consentRequest.consent.purpose",
                        "consent.purpose"
                    ],
                    "arguments": null,
                    "defaultMessage": "consent.purpose",
                    "code": "consent.purpose"
                }
            ],
            "defaultMessage": "Purpose is not specified.",
            "objectName": "consentRequest",
            "field": "consent.purpose",
            "rejectedValue": null,
            "bindingFailure": false,
            "code": "NotNull"
        },
        {
            "codes": [
                "NotNull.consentRequest.consent.callBackUrl",
                "NotNull.consent.callBackUrl",
                "NotNull.callBackUrl",
                "NotNull.java.lang.String",
                "NotNull"
            ],
            "arguments": [
                {
                    "codes": [
                        "consentRequest.consent.callBackUrl",
                        "consent.callBackUrl"
                    ],
                    "arguments": null,
                    "defaultMessage": "consent.callBackUrl",
                    "code": "consent.callBackUrl"
                }
            ],
            "defaultMessage": "Call back url not specified.",
            "objectName": "consentRequest",
            "field": "consent.callBackUrl",
            "rejectedValue": null,
            "bindingFailure": false,
            "code": "NotNull"
        }
    ]
}
```
This was a client failure, but spring returns 500, lets say we fix that somehow, but response body 
is completely different, client can be in any language, where they will look at the contract what we defined, and expect the response in that format at least for successful and client errors (4XX), either we fix our contract and return the response in same format for all client errors (business validation, unauthorized etc) or we fix our validation contract...

Reply:
I understand what you mean now. but I believe that can be handled either writing a custom error interpolation (or using spring bindingresult). The transformation of Javax binding result to this format, I believe is entirely possible.